nncf/common/quantization/quantizer_propagation/solver.py

```diff
@@ -0,0 +1,1211 @@
+            CatMetatype
+        ],
+        QuantizationTrait.OUTPUT_QUANTIZATION_AS_WEIGHTS: [
+            EmbeddingMetatype,
+            EmbeddingBagMetatype
+        ]
+    }  # type: Dict[QuantizationTrait, List[OperatorMetatype]]
+else:
+    DEFAULT_QUANT_TRAIT_TO_OP_DICT = {}  # TODO (vshampor): populate with TF metatypes or unify metatypes
+
+
+class TransitionStatus(Enum):
+    SHOULD_TRANSITION = 0
+    SHOULD_MERGE = 1
+    SHOULD_NOT_TRANSITION = 2
+    SHOULD_WAIT_FOR_MERGE = 3
+
+
+class PropagationStrategy(Enum):
+    # While propagating up through a downward-branching node:
+    # ... do not merge at all
+    DO_NOT_MERGE_BRANCH_FQS = 0
+
+    # ... only merge for exact configuration space matches
+    MERGE_IF_ALL_BRANCH_FQ_OPTIONS_SAME = 1
+
+    # ... merge common parts, and if a branch quantizer has options for
+    # narrowing (bitwidth/mode/per-channel/etc.) in addition to
+    # the common part, keep the quantizer on branch
+    MERGE_WITH_POTENTIAL_REQUANTIZATION = 2
+
+    # ... merge common config options into a single config space for the global FQ,
+    # do not merge if this is impossible for the current branching situation and given
+    # HW config file
+    MERGE_WITH_SINGLE_FQ_RESULT = 3
+
+
+class FinalizedQuantizationProposal:
+    def __init__(self, single_config_quantizer_setup: SingleConfigQuantizerSetup,
+                 quant_prop_graph: QuantizerPropagationStateGraph):
+        self.single_config_quantizer_setup = single_config_quantizer_setup
+        self._quant_prop_graph = quant_prop_graph
+
+    @property
+    def quant_prop_graph(self):
+        return self._quant_prop_graph
+
+
+class QuantizationProposal:
+    def __init__(self, quantizer_setup: 'MultiConfigQuantizerSetup',
+                 quant_prop_graph: QuantizerPropagationStateGraph,
+                 quantization_point_id_vs_prop_quantizer: Dict['QuantizationPointId',
+                                                               PropagatingQuantizer]):
+        self.quantizer_setup = quantizer_setup
+        self._quant_prop_graph = quant_prop_graph
+        self._quantization_point_id_vs_prop_quantizer = quantization_point_id_vs_prop_quantizer
+        self._prop_quantizer_vs_quantization_point_id = {}  # type: Dict[PropagatingQuantizer, QuantizationPointId]
+        for qp_id, pq in self._quantization_point_id_vs_prop_quantizer.items():
+            self._prop_quantizer_vs_quantization_point_id[pq] = qp_id
+
+    def constrain_quantizer_config_list_for_insertion(self, quantization_point_id: 'QuantizationPointId',
+                                                      constrained_config_list: List[QuantizerConfig]):
+        prior_list = self.quantizer_setup.quantization_points[quantization_point_id].possible_qconfigs
+        if not all([qc in prior_list for qc in constrained_config_list]):
+            raise RuntimeError("Constrained config list is incompatible with the result of the quantizer propagation!")
```

Comment:
Could you please use `'` instead of `"` here and below?

Reply:
Done